### ISSUE TITLE
Add a more memory-efficient RangeIndex-sort of thing to avoid large arange(N) indexes in some cases

### ISSUE THREAD

**Issue by wesm** [pydata/pandas#939]

**Comment by ARF1** [comment]
I am interested in RangeIndexes as well. Does anybody know what the state of @jtratner's tree is? Is there anything I can do to help push this towards a PR?

**Comment by jreback** [comment]
it's in a reasonable state
would welcome an update PR for it

**Comment by hayd** [comment]
I linked to it above, it's here https://github.com/jtratner/pandas/tree/add-range-index

IIRC it ought to rebase pretty cleanly (it's mainly new files). +1 to a resurrection!

**Comment by ARF1** [comment]
Ok, I rebased jtratner's tree: https://github.com/ARF1/pandas/tree/range_index

Of course it was too much to hope it would pass all the test. - One can always dream...

Having not worked on pandas internals before I am probably not the best person to move this forward efficiently. If somebody else feels like they are better suited I would be happy to contribute individual fixes to their tree.

If there are no takers, I will give it a shot myself. Though I will probably need a fair amount of hand-holding to get started.

Running the already written tests, it appears the `Index` api has changed. Is that possible? Can anybody point me to the PR / issue to help me understand how to adapt the code appropriately?

[pandas/core/range.py#L61-L76](https://github.com/ARF1/pandas/blob/a2c6ea35fdac8d262f777c9c92c53b9980270b59/pandas/core/range.py#L61-L76) has the following segment in its `__init__()` which seems no longer possible:
```
self = np.array([], dtype='int64').view(RangeIndex)
...
self.left = left
```
As I understand this, `RangeIndex` being a subclass of `Index` used to be a `ndarray` subclass but is no longer, right?

**Comment by jreback** [comment]
yes Index is no longer a subclass of ndarray since 0.15.0.You can look at the recently merged CategoricalIndex for the api

**Comment by jorisvandenbossche** [comment]
@ARF1 The easiest thing to do is probably to just open a pull request based on your branch. Then it is easier to comment on the code and give feedback.

**Comment by ARF1** [comment]
@jorisvandenbossche Ok, I did not want to "pollute" the PR list with unfinished code. The rebased tree is available as PR #9961. 

@jreback Thanks. I will take a look and see if I can find my way around...

**Comment by ARF1** [comment]
Ok, the move away from ndarray Indexes has quite a few knock-on effects on jtratner's tree. I am thinking I need to rewrite the test suite first.

Would it make sense to copy and adapt the test suite of the `Int64Index` and progressively eliminate the test failures? My thought was that `RangeIndex` should behave largely identical viewed from the outside (apart from instantiation).

**Comment by jreback** [comment]
yes it should be a drop in replacement (and inherit from Int64Index)

**Comment by shoyer** [comment]
The CategoricalIndex PR (https://github.com/pydata/pandas/pull/9741) is actually a really nice example of how to make this work.

**Comment by ARF1** [comment]
A working (at least apparently so) implementation of `RangeIndex()` is now available as PR #9977.

I decided to re-implement rather than use existing trees: the change in the `Index` api was making things too difficult to adapt.

This implementation is fully compatible to `Int64Index()` and automatically falls back to its implementation when required. If this happens, an equivalent `Int64Index()` is created on-the-fly with obvious performance implications.

At present only with a limited set of operations the `RangeIndex()` is conserved: e.g. slicing, inner-, left- and right-joins, `size` and `__len__()`, .

My aim would be to merge with the current feature set as soon as possible and add more type-conserving operations with subsequent PRs to avoid the fate of the previous efforts with `RangeIndex()` which seem to have been overtaken by the continuing development of pandas.

I would welcome any advice and comments for making this happen.

Special thanks to go @jreback and @shoyer for their advice and helping me get started.
